packages/react-client/src/devices/MiddlewareManager.ts

```diff
@@ -0,0 +1,77 @@
+import type { Media } from "../types/internal";
+
+import type { TrackMiddleware } from "../types/public";
+import { setupOnEndedCallback } from "../utils/track";
+
+type StreamTrack = Pick<Media, "track" | "stream" | "enabled">;
+
+export class MiddlewareManager {
+  private media: StreamTrack | null = null;
+
+  private middleware: TrackMiddleware | null = null;
+  private clearMiddlewareFn: (() => void) | null = null;
+
+  public getMedia = (): StreamTrack | null => this.media;
+
+  public updateMedia(media: StreamTrack | null) {
+    this.media = !media ? null : this.applyMiddleware(media.track, this.middleware);
+  }
+
+  public setTrackMiddleware(middleware: TrackMiddleware | null, rawTrack: MediaStreamTrack | null) {
+    this.middleware = middleware;
+    this.clearMiddleware();
+
+    if (!rawTrack) return;
+
+    this.media = this.applyMiddleware(rawTrack, middleware);
+  }
+
+  public getMiddleware(): TrackMiddleware | null {
+    return this.middleware;
+  }
+
+  public stop() {
+    this.clearMiddleware();
+    this.media?.track?.stop();
+  }
+
+  public disable() {
+    if (!this.media?.track) return;
+
+    this.media.track.enabled = false;
+    this.media.enabled = false;
+  }
+
+  public enable() {
+    if (!this.media?.track) return;
+
+    this.media!.track!.enabled = true;
+    this.media!.enabled = true;
+  }
```

Comment:
Ad 2. There are two distinct pairs of Stream + Track. One is the raw stream from a device, while the second is the result of the raw stream processed by middleware. Stopping the raw stream probably doesn't stop the middleware stream, because the latter is purely virtual, and the middleware is responsible for updating the stream